proposed/0000-crust-refactor/0000-crust-refactor.md

```diff
@@ -0,0 +1,501 @@
+                            ::NOTIFY_WAKE => {
+                                if inner.wake_buffer > 0 {
+                                    inner.wake_buffer -= 1;
+                                }
+                                else {
+                                    inner.fresh = false;
+                                    while !inner.fresh {
+                                        inner.in_accept = true;
+                                        inner = self.acceptor.inner_freshened.wait(inner).unwrap();
+                                        inner.in_accept = false;
+                                    }
+                                }
+                            },
+                            ::NOTIFY_SHUTDOWN => return None,
+                            x => panic!("Unexpected byte in notify pipe: {}", x),
+                        }
+                    },
+                    t => {
+                        match inner.listeners[t.0].accept() {
+                            Ok(None)         => (),
+                            Ok(Some(stream)) => return Some((self, Ok(Stream::Tcp(stream)))),
+                            Err(e)           => return Some((self, Err(AcceptError::Accept {cause: e}))),
+                        }
+                    },
+                }
+            }
+        }
+    }
+
+    fn go_opt_timeout(self, opt_timeout: Option<Duration>) -> Option<(AcceptGo<'a>, Result<Stream, AcceptError>)> {
+        self.go_timeout(opt_timeout.unwrap_or(Duration::max_value()))
+    }
+
+    fn go_opt_deadline(self, opt_deadline: Option<SteadyTime>) -> Option<(AcceptGo<'a>, Result<Stream, AcceptError>)> {
+        self.go_timeout(opt_deadline.map(|d| d - SteadyTime::now()).unwrap_or(Duration::max_value()))
+    }
+}
+
+pub struct AcceptorController<'a> {
+    acceptor: &'a Acceptor,
+}
+
+impl<'a> AcceptorController<'a> {
+    pub fn add_listener<A: ToListenEndpoint>(&self, addr: A) -> Result<(), AddListenerError<A::Err>> {
+        let addr = try!(addr.to_listen_endpoint()
+                             .map_err(|e| AddListenerError::ParseAddr {cause: e}));
+        match addr {
+            ListenEndpoint::Tcp(sa) => {
+                let listener = try!(TcpListener::bind(&sa)
+                                                .map_err(|e| AddListenerError::Bind {cause: e}));
+                {
+                    let mut notify_write = self.acceptor.notify_write.lock().unwrap();
+                    try!(notify_write.write_all(&[::NOTIFY_WAKE])
+                                     .map_err(|e| AddListenerError::Notify {cause: e}));
+                }
+                {
+                    let mut inner = self.acceptor.inner.lock().unwrap();
+                    let token = mio::Token(inner.listeners.len());
+                    try!(inner.poll.register(&listener, token, mio::EventSet::readable(), mio::PollOpt::level())
+                                   .map_err(|e| AddListenerError::Register {cause: e}));
```

Comment:
Because `AcceptGo::go` returns after accepting one connection. If we receive two connections consecutively then we want to be notified about the second.